File: clients/typescript/src/cache_watcher.ts
import { ICacheWatcher } from './cache_manager';
import { connectionManager } from './connection_manager';
import { AmenoException } from './shared';

export type CacheAllHandler = (key: string, value: any) => void;
export type CacheKeyHandler<T> = (value: T | null) => void;

interface KeySubscription {
    handler: (rawValue: any) => void;
}

export class CacheWatcher implements ICacheWatcher {
    public readonly group: string;
    private disposed = false;
    private readonly allHandlers: Set<CacheAllHandler> = new Set();
    private readonly keyHandlers: Map<string, Set<KeySubscription>> = new Map();

    constructor(group: string) {
        this.group = group;
    }

    public all(handler: CacheAllHandler): void {
        this.ensureNotDisposed();
        this.allHandlers.add(handler);
        connectionManager.cacheManager.subscribeWatcher(this);
    }

    public key<T>(key: string, handler: CacheKeyHandler<T>): void {
        this.ensureNotDisposed();
        if (!this.keyHandlers.has(key))
            this.keyHandlers.set(key, new Set());

        const subscriptions = this.keyHandlers.get(key)!;
        const wrapper = (rawValue: any): void => {
            handler(rawValue as T | null);
        };

        subscriptions.add({ handler: wrapper });
        connectionManager.cacheManager.subscribeWatcher(this);
    }

    public dispose(): void {
        this.ensureNotDisposed();
        this.disposed = true;
        this.allHandlers.clear();
        this.keyHandlers.clear();
        connectionManager.cacheManager.unsubscribeWatcher(this);
    }

    public dispatchValueChanged(key: string, rawValue: any): void {
        if (this.disposed)
            return;

        for (const handler of Array.from(this.allHandlers))
            handler(key, rawValue);

        const subscriptions = this.keyHandlers.get(key);
        if (subscriptions)
            for (const subscription of Array.from(subscriptions))
                subscription.handler(rawValue);
    }

    private ensureNotDisposed(): void {
        if (this.disposed)
            throw new AmenoException(`O observador do grupo de cache '${this.group}' já foi descartado (disposed).`);
    }
}

File: clients/csharp/AmenoLink/CacheWatcher.cs
namespace AmenoLink;

public class CacheWatcher(string group) : ICacheWatcher, ICacheWatcherSubscriber
{
    public string Group { get; } = group;
    private bool disposed;
    private readonly HashSet<Action<string, object?>> allHandlers = [];
    private readonly Dictionary<string, List<Action<object?>>> keyHandlers = [];
    private readonly Lock lockObject = new();

    public void All(Action<string, object?> handler)
    {
        EnsureNotDisposed();
        lock (lockObject)
            allHandlers.Add(handler);

        AmenoLinkClient.ConnectionManager.CacheManager.SubscribeWatcher(this);
    }

    public void Key<T>(string key, Action<T?> handler)
    {
        EnsureNotDisposed();
        lock (lockObject)
        {
            if (!keyHandlers.TryGetValue(key, out var list))
            {
                list = [];
                keyHandlers[key] = list;
            }

            list.Add(rawValue =>
            {
                if (rawValue is null)
                {
                    handler(default);
                    return;
                }

                if (rawValue is T typed)
                {
                    handler(typed);
                    return;
                }

                var serialized = System.Text.Json.JsonSerializer.Serialize(rawValue, AmenoLinkClient.JsonOptions);
                var deserialized = System.Text.Json.JsonSerializer.Deserialize<T>(serialized, AmenoLinkClient.JsonOptions);
                handler(deserialized);
            });
        }

        AmenoLinkClient.ConnectionManager.CacheManager.SubscribeWatcher(this);
    }

    public void Dispose()
    {
        EnsureNotDisposed();
        disposed = true;

        lock (lockObject)
        {
            allHandlers.Clear();
            keyHandlers.Clear();
        }

        AmenoLinkClient.ConnectionManager.CacheManager.UnsubscribeWatcher(this);
        GC.SuppressFinalize(this);
    }

    void ICacheWatcherSubscriber.DispatchValueChanged(string key, object? value)
    {
        if (disposed)
            return;

        List<Action<string, object?>> allTargets;
        List<Action<object?>> keyTargets = [];

        lock (lockObject)
        {
            allTargets = [.. allHandlers];
            if (keyHandlers.TryGetValue(key, out var list))
                keyTargets = [.. list];
        }

        foreach (var handler in allTargets)
            handler(key, value);

        foreach (var handler in keyTargets)
            handler(value);
    }

    private void EnsureNotDisposed()
    {
        if (disposed)
            throw new AmenoException($"O observador do grupo de cache '{Group}' já foi descartado (disposed).");
    }
}
